helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java

```diff
@@ -487,10 +490,18 @@ private void chargePendingTransition(Resource resource, CurrentStateOutput curre
           String pendingState = pendingMap.get(instance);
           if (pendingState != null && !pendingState.equals(currentState)
               && !cache.getDisabledInstancesForPartition(resourceName, partition.getPartitionName())
-                  .contains(instance)) {
+              .contains(instance)) {
             // Only charge this instance if the partition is not disabled
             throttleController.chargeInstance(rebalanceType, instance);
             shouldChargePartition = true;
+            // Remove the partition marked as partition need recovery/load balance. Otherwise, they
+            // will be double charged when they already charged for pending messages.
+            // At the same time, we need to update the intermediate map since it was supposed to
+            // be updated when the recovery rebalance throttling piece.
+            cleanPartitionsAreCharged(partition, partitionsNeedRecovery, partitionsNeedLoadbalance,
+                rebalanceType);
+            intermediatePartitionStateMap
```

Comment:
bestPossiblePartitionStateMap and intermediatePartitionStateMap updates are addition to the bug fix, right? Why is it needed?

Reply:
This is used for assignment. Intermediate assign should be null at first beginning. It will be filled with either BestPossible or CurrentState when it has been throttled or not.

Reply:
Or maybe put it in another question: what if we don't set the intermediatePartitionStateMap after the fix, what's wrong with it? Or shall we add the update logic if we don't have the fix?